### PR TITLE
ARROW-14071: [R] Try to arrow_eval user-defined functions - explicit approach

### PR DESCRIPTION
**Pull request by dragosmg**
Once this PR is merged, users should be able to use their own functions in an {arrow} dplyr-like pipeline (provided they can be translated with the help of existing bindings). 

Thoughts around the design were captured in this [design doc](https://docs.google.com/document/d/1vBp8M8yXXfwMfpAPTXPKVT6QvhGrZWc0Uv5_xpEOq00/edit#).

This PR implements the _explicit_ approach, while #13960 implements the _implicit_ one.
* _explicit_: walks through the abstract syntax tree and translates each _translatable_ function call encountered
    - works with `filter()`, `mutate()` and `transmute()` , i.e. `nse_funcs`
    - doesn’t yet work with `summarise()`, i.e. `agg_funcs`
    - the code is a bit more complex and might require more effort to maintain
    - nested translations work
* _implicit_: modifying the environment of the unknown functions
  - more difficult to get nested translations right (doing the registration recursively), but they work
  - works with `filter()`, `mutate()` and `transmute()`
  - doesn’t yet work with `summarise()`
  - less complex in terms of code, but more complex in terms of mental model (getting recursion right when environments are involved is a bit trickier), but potentially easier to maintain.

<details>
  <summary>Examples</summary>

``` r
library(dplyr)
library(arrow)

nchar2 <- function(x) {
  1 + nchar(x)
}

# simple expression
tibble::tibble(my_string = "1234") %>%
  mutate(
    var1 = nchar(my_string),
    var2 = nchar2(my_string)) %>%
  collect()
#> # A tibble: 1 × 3
#>   my_string  var1  var2
#>   <chr>     <int> <dbl>
#> 1 1234          4     5

# a slightly more complicated expression
tibble::tibble(my_string = "1234") %>%
  mutate(
    var1 = nchar(my_string),
    var2 = 1 + nchar2(my_string)) %>%
  collect()
#> # A tibble: 1 × 3
#>   my_string  var1  var2
#>   <chr>     <int> <dbl>
#> 1 1234          4     6
  
nchar3 <- function(x) {
  2 + nchar(x)
}

# multiple unknown calls in the same expression (to test the iteration)
tibble::tibble(my_string = "1234") %>%
  mutate(
    var1 = nchar(my_string),
    var2 = nchar2(my_string) + nchar3(my_string)) %>%
  collect()
#> # A tibble: 1 × 3
#>   my_string  var1  var2
#>   <chr>     <int> <dbl>
#> 1 1234          4    11

# user function defined using namespacing
nchar4 <- function(x) {
  2 + base::nchar(x)
}

tibble::tibble(my_string = "1234") %>%
  mutate(
    var1 = nchar(my_string),
    var2 = 1 + nchar4(my_string)) %>%
  collect()
#> # A tibble: 1 × 3
#>   my_string  var1  var2
#>   <chr>     <int> <dbl>
#> 1 1234          4     7

isPowerfulEngine <- function(x) x > 200
mtcars %>% 
  arrow_table() %>% 
  filter(isPowerfulEngine(hp)) %>% 
  collect()

#>    mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> 1 14.3   8  360 245 3.21 3.570 15.84  0  0    3    4
#> 2 10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> 3 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4
#> 4 14.7   8  440 230 3.23 5.345 17.42  0  0    3    4
#> 5 13.3   8  350 245 3.73 3.840 15.41  0  0    3    4
#> 6 15.8   8  351 264 4.22 3.170 14.50  0  1    5    4
#> 7 15.0   8  301 335 3.54 3.570 14.60  0  1    5    8
```

</details>

### CODE REVIEW

**Comment by jonkeane** [comment]
I'm not wholly against requiring registration if it is significantly simpler or is the only way. But I'm not sure that requiring registration will necessarily help make errors easier to understand (we'll need to have some good, clear error messages when anything goes wrong here regardless of the path we take). 

And I might even argue that having only one kind of registration (for the user-defined scalar function) sets that apart in a positive way (crafting one of those will take care + do things that are surprising compared to all of our other bindings, after all). IMO: people shouldn't (need to) think too much about how the bindings work, just use the code you're used to, and if it works, great, if it gives you a (clear) error that some function isn't native to arrow yet, you can respond to that as is. If we have two different types of registration we'll need to find a way to explain how they are different (and even with great docs explaining that difference, there will be a contingent of people who will wonder what's the difference between the hypothetical `register_binding(...)` and `register_scalar_function(...)`.

But, again, if autoregistering is a sticking point (though looking through the code that's here it doesn't look like it is...), then sure we can go the registration route first. (or if all the autoregistration stuff is overly burdensome from a maintenance perspective too, of course).

File: r/R/dplyr-eval.R [hunk 1]
```diff
@@ -121,3 +148,121 @@ format_expr <- function(x) {
...
+      body = translate_to_arrow_rec(function_body[[2]])
+    )
+  } else {
+    # TODO WIP if the function is not directly translatable, make one more try
+    # by attempting to translate the unknown calls
+    unknown_function <- setdiff(all_funs(function_body[[2]]), names(.env))
+
+    fn <- as_function(unknown_function, env = caller_env())
+    translated_function <- NULL
+  }
```
Comment (jonkeane) [inline r/R/dplyr-eval.R:188]: If I'm reading this code correctly, it looks like it translates two levels down in the function call, yeah? So if I have only functions in my function body that have bindings it'll be fine, or if my function calls a function that also has only functions that have bindings it'll be fine, but if my function calls a function, which calls another function, which has only bindings it won't work.

In other words: these will work:
```
my_func <- function(x) {
  mean(x)
}

my_func_2 <- function(x) {
  my_func(x)
}
```

but this one will not:
```
my_func_3 <- function(x) {
  my_func_3(x)
}
```

Or does it actually recurse through the call stack?

**Review by jonkeane — COMMENTED** [review]
Thanks for this! One comment, so far. It would be nice to have an idea of which parts of this are working, and which parts still need work before they are ready (I see a handful of todos in the code, which is great, but also in a comment of "I've done these bits: ..., still working on these: ...")

File: r/R/dplyr-eval.R (hunk 1 above)
Comment (dragosmg) [reply r/R/dplyr-eval.R:188]: It doesn't recurse through the call stack at the moment. Still working on it and I will implement it as recursing only 1 level down in the first instance.

**Comment by dragosmg** [comment]
> Thanks for this! One comment, so far. It would be nice to have an idea of which parts of this are working, and which parts still need work before they are ready (I see a handful of todos in the code, which is great, but also in a comment of "I've done these bits: ..., still working on these: ...")

All the examples listed in the PR description work. So far this works for:
* mutate with:
    - [ ] simple expression (`nchar2(my_string)`) - contains a simple call to a user-defined function
    - [ ] a slightly more complicated expression (`1 + nchar2(my_string)`)
    - [ ] multiple unknown calls in the same expression (to test the iteration) - `nchar2(my_string) + nchar3(my_string)`
    - [ ] user function defined using namespacing `1 + nchar4(my_string)`

**Comment by dragosmg** [comment]
@paleolimbot I like your suggestion in https://github.com/apache/arrow/pull/13789#issuecomment-1214992856. Even if we don't require the user to self-register, that function can really simplify my current logic (as we can use it instead of having to walk the AST).

File: r/R/dplyr-eval.R (hunk 1 above)
Comment (dragosmg) [reply r/R/dplyr-eval.R]: This now works, and will mark this as resolved. There's a lot of polishing / simplification I want to do.

**Comment by dragosmg** [comment]
**Update**:
* user-defined functions work with `mutate()` and `filter()`
* and work with nested function definitions

**Comment by dragosmg** [comment]
> > Give both of these work and the approach is rather complex, by walking the AST, I will now go more in the direction suggested by @paleolimbot (in [#13789 (comment)](https://github.com/apache/arrow/pull/13789#issuecomment-1214992856)) and refactor / simplify the implementation.
> 
> Could you say more about what you're thinking about doing here, there were a few suggestions in that comment | thread. Are you going to now require registration? Or something else?

Not requiring registration. I was thinking about switching from translation + registration to registration only.

Currently my approach is to take:
```r
nchar2 <- function(x) 1 + nchar(x)
```
walk its abstract syntax tree, translate it to 
```r
nchar2 <- function(x) call_binding("+", 1, call_binding("char", x))
```
and then register it + update `.cache$functions`.

However, the translation might not be necessary and only complicates the implementation, as we only need to register the `"nchar2"` binding without actually translating its body for this to work. The reason it doesn't work in `main` is because the data mask has no knowledge of `"nchar2"`. So updating the context of the evaluation (to be aware of the _translation_, but without actually translating the body) should work.

Translating the body also highlighted the overlap between `nse_funcs` and `.cache$functions`. We could potentially reduce that complexity too - [ARROW-17495](https://issues.apache.org/jira/browse/ARROW-17495), but I'm happy to leave that for another PR.

**Comment by jonkeane** [comment]
Ah cool, that sounds good, thanks!

**Comment by dragosmg** [comment]
The CI failure (Windows & R 3.6) seems to be related to https://github.com/apache/arrow/pull/13661.

**Comment by thisisnic** [comment]
I'm closing this PR for now as it was handed over to me to take a look at, but I don't think I'm going to end up with time to work on this in the near future, and we're just having a bit of a tidy of PRs ahead of the upcoming release.

**Pull request closed without merging**